common/lib/xmodule/xmodule/modulestore/django.py

```diff
@@ -243,10 +245,47 @@ class ModuleI18nService(object):
     i18n service.
 
     """
+    def __init__(self, block):
+        """
+        Store a reference to the block currently being serviced by this runtime.  We'll use this
+        reference later on for things like locating the block's translation domain (PO+MO files)
+        """
+        self.block = block
 
     def __getattr__(self, name):
         return getattr(django.utils.translation, name)
 
+    def ugettext(self, string):
+        """
+        This operation is a proxy for django.utils.translation.ugettext, which is itself a proxy for
+        the GNU gettext ugettext operation.  Here we attempts to look up the provided string in the
+        XBlock's own domain translation catalog, currently expected to be:
+            <xblock_root>/conf/locale/<language>/LC_MESSAGES/<domain>.po
+        If ModuleI18nService can't locate the domain translation catalog then we fall-back onto
+        django.utils.translation.ugettext, which will attempt to find a matching string in the
+        LMS' own domain translation catalog -- effectively achieving translation by coincidence.
+        """
+        # The translation workflow should only execute if there's an actual string to look up
+        # If gettext processes an empty string the PO file header information will oddly be returned
+        translated_string = unicode(string)
+        if translated_string and self.block:
+            try:
+                xblock_resource = self.block.__class__.unmixed_class.__module__
+                xblock_locale_dir = '/conf/locale'
+                xblock_locale_path = resource_filename(xblock_resource, xblock_locale_dir)
+                xblock_domain = 'django'
+                selected_language = get_language()
+                translator = gettext.translation(
+                    xblock_domain,
+                    xblock_locale_path,
+                    [to_locale(selected_language if selected_language else settings.LANGUAGE_CODE)]
+                )
```

Comment:
@asadiqbal08 all of the work on lines 273-282 is the same no matter what string is being requested. It doesn't make sense to do it every time a string is translated. It should be done only once. It's really unlikely an XBlock will ask to translate an empty string, and it's especially unlikely that they'll 1) ask to translate an empty string, and 2) never ask to translate another string, and 3) mind having done this work.

This should be moved into a place that it is done only once.  Also, remember that we're going to have to provide an implementation of ungettext as well.  We should get that in place so that we can start making the right decisions for the multiple methods we'll have to support.